### PR TITLE
clarify explanation of map_values, including description of new behavior

### PR DESCRIPTION
**Pull request by pkoppstein**
#2466

### CODE REVIEW

File: docs/content/manual/manual.yml [hunk 1]
```diff
@@ -925,6 +929,15 @@ sections:
             input: '{"a": 1, "b": 2, "c": 3}'
             output: ['{"a": 2, "b": 3, "c": 4}']
 
+          - program: 'map(., .)'
+            input: '[1,2]'
+            output: ['[1,1,2,2]']
+
+          - program: 'map_values(. // empty)'
+            input: '{"a": null, "b": true, "c": false}'
+            output: ['{"b":true}']
```
Comment (wader) [inline docs/content/manual/manual.yml:967]: Too much to have object with `map` and array with `map_values` examples?

**Review by wader — APPROVED** [review]

File: docs/content/manual/manual.yml [hunk 2]
```diff
@@ -908,14 +908,18 @@ sections:
...
+          array. For example, `map(.+1)` will increment each element
+          of an array of numbers, and `map(.,.)` will duplicate each
+          element of the input array.
 
-          `map(f)` is equivalent to `[.[] | f]`. In fact, this is how
-          it's defined. Similarly, `map_values(f)` is defined as `.[] |= f`.
+          `map(f)` is equivalent to `[.[] | f]`. In fact, this is how it's defined.
 
+          `map_values(f)` is similarly defined as `.[] |= f`.  This
+          essentially means that each value, $v, in the input array or
```
Comment (nicowilliams) [inline docs/content/manual/manual.yml]: `s/, [$]v,//`

File: docs/content/manual/manual.yml [hunk 3]
```diff
@@ -908,14 +908,18 @@ sections:
...
+          object is replaced by `first($v|f)`, it being understood
+          that if the input is an object, then keys for which this
```
Comment (nicowilliams) [inline docs/content/manual/manual.yml]: Array values for which `f` returns `empty` also get deleted.

I liked the original text better.  Staring at the definitions of `map/1` and `map_values/1` the difference between them isn't quite obvious, but the original text here makes it obvious: `map/1` always returns an array, while `map_values/1` returns an array if its input was an array, or an object if its input was an object.

Comment (pkoppstein) [reply docs/content/manual/manual.yml]: Tx. Hopefully the subsequent commit addresses all these concerns.  It is more systematic about the various possibilities.

File: docs/content/manual/manual.yml [hunk 4]
```diff
@@ -906,15 +906,42 @@ sections:
...
-          For any filter `f`, `map(f)` will run that filter for each
-          element of the input array, and return the outputs in a new
-          array. `map(.+1)` will increment each element of an array of numbers.
+          For any filter `f`, `map(f)` and `map_values(f)` apply `f`
+          to each of the values in the input array or object, that is,
+          to the values of `.[]`.
+          
+          In the absence of errors, `map(f)` always outputs an array
+          whereas `map_values(f)` outputs an array or an object
+          depending on the type of the input.
```
Comment (nicowilliams) [inline docs/content/manual/manual.yml]: `In the absence of errors, ` is not needed.

Can you add something about the objects' keys being preserved?  Something like:

```
          `map(f)` always outputs an array whereas `map_values(f)` outputs
          an array if given an array, or an object if given an object.  In the
          case of `map_values(f)` applied to an object, the object output
          has all the same keys as the input object minus any for whose
          values `f` produced no new values.
```

something like that.

Comment (pkoppstein) [reply docs/content/manual/manual.yml]: Done. Tx.

**Comment by nicowilliams** [comment]
Thanks!

**Pull request merged**